rust/bindings/build.rs

```diff
@@ -0,0 +1,260 @@
+//! This crate interprets all '*.json' files in `/specs` as OpenRPC documents,
+//! and generates corresponding modules based on the filename.
+
+use anyhow::{bail, Context as _};
+use proc_macro2::TokenStream;
+use quote::quote;
+use schemars::schema::{
+    ArrayValidation, ObjectValidation, RootSchema, Schema, SchemaObject, SingleOrVec,
+    SubschemaValidation,
+};
+use syn::Ident;
+
+use std::{
+    env,
+    ffi::OsStr,
+    fs::{self, File},
+    path::Path,
+};
+
+fn main() -> anyhow::Result<()> {
+    let dir = concat!(env!("CARGO_MANIFEST_DIR"), "/specs/");
+    println!("cargo::rerun-if-changed={}", dir);
+
+    let mut modules = TokenStream::new();
+    for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        if entry.metadata()?.is_file()
+            && path.extension().is_some_and(|it| it == "json")
+            && path
+                .file_name()
+                .is_some_and(|it| !it.to_string_lossy().starts_with('.'))
+        {
+            modules.extend(
+                file2module(&path).context(format!("couldn't process file {}", path.display()))?,
+            )
+        }
+    }
+
+    let pretty = prettyplease::unparse(&syn::parse2(modules).context("internal codegen error")?);
+    let mut out = env::var_os("OUT_DIR").context("invalid build script environment")?;
+    out.push("/bindings.rs");
+    fs::write(out, pretty).context("couldn't write generated code")?;
+
+    Ok(())
+}
+
+fn file2module(path: &Path) -> anyhow::Result<TokenStream> {
```

Comment:
nit: let's have comments for each method in this file to have an easier time reading, understanding and potentially maintaining this code.